Fix BeatV2Manager to configure inputs and set log level

blakerouse opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

This refactors the `BeatV2Manager` so it works correctly with the Elastic Agent V2 model of components/units. The log level is now computed from the defined units and the log level is now updated with the `logp.SetLevel`. 

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

Previously the `BeatV2Manager` would act incorrectly when a new unit was added to the beat from the Elastic Agent replacing the already existing unit with the new unit configuration instead of merging the configuration so the beat would operating with 2 inputs. Previous the log level was not set-able by the V2 control protocol, this now works.
## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes #34048 
- Closes #34046
- Closes https://github.com/elastic/elastic-agent/issues/1913
- Closes #34093

blakerouse commented:
I am keeping this in draft at the moment unit I get the unit tests done, but I have tested this manually with the Elastic Agent and it is working correctly. Still not comfortable to land this until I have unit tests to cover it. Wanted to get it up early so others can review and test it as well.

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34066/detail/PR-34066/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34066/detail/PR-34066/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34066/detail/PR-34066/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34066/detail/PR-34066/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-12-16T13:30:52.435Z&rangeTo=2022-12-16T13:50:52.435Z&transactionName=BUILD+Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=3e0b8a113c3ed1e17fbbaf39037d9085&transactionId=f25b33cd08aa4bc9)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-12-16T13:40:52.435+0000


* Duration: 6 min 53 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34066/detail/PR-34066/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-34066/runs/1/steps/106/log/?start=0">here</a></li>
<li>Description: <code>tar: step failed with error null</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

fearful-symmetry commented on x-pack/libbeat/management/managerV2.go:
```diff
@@ -311,95 +359,162 @@ func (cm *BeatV2Manager) stopBeat() {
+			// `reload` method and will be marked stopped in that code path)
+			continue
+		}
+		err := unit.UpdateState(status, message, payload)
```
Not sure why we're manually looping over the unit `UpdateState` here? Is there a reason why we just can't call the global `UpdateStatus` or something similar?

blakerouse replied:
The global `UpdateStatus` will grab the mutex for units, to prevent that we call this logic on the units that reload just performed. We also would not want to update the status of units that might have been added to the `cm.units` that is different than the actual units that are being processed in the reload (because they where passed in).

fearful-symmetry commented on x-pack/libbeat/management/managerV2.go:
```diff
@@ -251,38 +315,29 @@ func (cm *BeatV2Manager) unitListen() {
-				cm.logger.Debugf("Got unit removed: %s", change.Unit.ID())
 				cm.deleteUnit(change.Unit)
 			}
+		case <-cm.reloadCh:
```
I'm not sure why there's so much logic involved in the reloading process? `triggerReload()` and `reload()` are both only called in this for block, it might just be simpler to call `reload()` in a goroutine or something? Also, why is `reload()` getting its own map of the units and not just using `cm.units`?

blakerouse replied:
This is because performing `reload` might result in the config logic of the beat to call `UpdateStatus` which will then grab the mutex for the units.

So to ensure that a dead lock doesn't occur we ensure that the actual reload logic is performed in the main loop of the manager, but not in a path that would hold that mutex. So that is why a copy is sent to the reload function. Each `client.Unit` has an internal lock for state so it is also save to have a pointer to the same unit.

fearful-symmetry replied:
When you mention a deadlock starting with `UpdateStatus` I assume you mean the `Reload()` call blocking while the beat waits for `UpdateStatus`? 

I wonder if we can do something like have a `addUnit()` return a new copy of the unit map that's sent to `reload()` so we can avoid the extra `reloadCh`? If not, can you at least add a comment explaining the loop, was a tad confusing at first.

cmacknz replied:
Can you add a comment explaining the potential for deadlock here so it's obvious the next time we read this code?

blakerouse replied:
I have added a comment as well as add some debounce to consolidate unit changes into a single reload. It makes it simpler to understand as well.

fearful-symmetry reviewed (commented):
A few comments/questions, mostly about overall logic.

blakerouse commented on x-pack/libbeat/management/managerV2.go:
```diff
@@ -23,6 +25,11 @@ import (
 	"github.com/elastic/elastic-agent-libs/logp"
 )
 
+type unitKey struct {
```
Only the Type/ID is unique. It is possible that an input ID and and output ID could be the same. I will add a comment.

blakerouse commented on x-pack/libbeat/management/managerV2.go:
```diff
@@ -31,15 +38,20 @@ type BeatV2Manager struct {
-	units    map[string]*client.Unit
-	mainUnit string
+	// track individual units given to us by the V2 API
+	mx      sync.Mutex
```
Yeah I can move it. I originally have multiple mutex for different parts but then I would still need to grab the unitsMx so I just changed it to a single mutex because it was safer than causing some deadlock in lock ordering.

blakerouse commented on x-pack/libbeat/management/managerV2.go:
```diff
@@ -158,69 +171,120 @@ func (cm *BeatV2Manager) CheckRawConfig(_ *conf.C) error {
+	defer cm.mx.Unlock()
+
+	cm.actions = append(cm.actions, action)
+	for _, unit := range cm.units {
```
Technically they shouldn't be registered in every unit, it really should be a specific unit thing. But the problem is that beats doesn't pass the unit down to the implementation to be able to make that type of registration and be specific.

So the actions must be registered on every unit that way the action will actually be called. There is no such thing as a main unit or anything like that, so all must get the action.

fearful-symmetry commented on x-pack/libbeat/management/managerV2.go:
```diff
@@ -311,95 +359,162 @@ func (cm *BeatV2Manager) stopBeat() {
+	// now update the statuses of all units
+	cm.mx.Lock()
+	status := getUnitState(cm.status)
+	message := cm.message
```
What's the reasoning behind using the global message? Seems like we should have a status message specifically mentioning the reload operation?

blakerouse replied:
This is basically setting the units to Healthy, or what ever beats has set with the `UpdateStatus` for the current state of the beat. The state is global because beats doesn't have a way of passing each unit to each running input and have it handle its own state.

blakerouse commented on x-pack/libbeat/management/managerV2.go:
```diff
@@ -158,69 +171,120 @@ func (cm *BeatV2Manager) CheckRawConfig(_ *conf.C) error {
-		cm.mainUnit = unit.ID()
+func (cm *BeatV2Manager) modifyUnit(unit *client.Unit) {
+	// `unit` is already in `cm.units` no need to add it to the map again
+	// but the lock still needs to be help so reload can be triggered
```
Fixed

blakerouse commented:
Okay I think this is ready to be merged. I have added unit tests, answered all questions, updated comments, an added debounce to unit changes. I also merged https://github.com/elastic/beats/pull/34049 into this PR so its already present and ready to go once this lands.

cmacknz commented on x-pack/libbeat/management/managerV2.go:
```diff
@@ -31,26 +42,33 @@ type BeatV2Manager struct {
-	// This satisfies the SetPayload() function, and will pass along this value to the UpdateStatus()
-	// call whenever a config is re-registered
+	// status is reported as a whole for every unit sent to this component
+	// hopefully this can be improved in the future to be seperated per unit
```
```suggestion
	// hopefully this can be improved in the future to be separated per unit
```

cmacknz commented on x-pack/libbeat/management/managerV2.go:
```diff
@@ -23,6 +27,13 @@ import (
+// the `ID` of a unit in itself is not unique without its type, only `Type` + `ID` is unique
+type unitKey struct {
+	Type client.UnitType
+	ID   string
```
Is there anything on the agent side enforcing that this tuple is unique per unit?

Trying to trace where this is coming from and I see this on the agent side: https://github.com/elastic/elastic-agent/blob/7e376c3eec28d0bbd9064930042a2225e9a29dbc/pkg/component/component.go#L235-L241

```go
				units = append(units, Unit{
					ID:       fmt.Sprintf("%s-%s-%s", inputType, outputName, input.id),
					Type:     client.UnitTypeInput,
					LogLevel: input.logLevel,
					Config:   cfg,
					Err:      cfgErr,
				})
```

The uniqueness is coming from the `input.id` primarily, which I see parsed out of the agent policy with a uniqueness check so it seems to be there: https://github.com/elastic/elastic-agent/blob/7e376c3eec28d0bbd9064930042a2225e9a29dbc/pkg/component/component.go#L511-L522

```go
		idRaw, ok := input[idKey]
		if !ok {
			// no ID; fallback to type
			idRaw = t
		}
		id, ok := idRaw.(string)
		if !ok {
			return nil, fmt.Errorf("invalid 'inputs.%d.id', expected a string not a %T", idx, idRaw)
		}
		if hasDuplicate(outputsMap, id) {
			return nil, fmt.Errorf("invalid 'inputs.%d.id', has a duplicate id %q (id is required to be unique)", idx, id)
		}
```

Was this uniqueness check there before? Does it work with standalone agent policies that omit the ID field? We don't seem to document the need for it in our standalone agent docs so likely we will encounter agent policies with missing input IDs in the wild, possibly lots of them.

blakerouse replied:
ID is a required value on standalone agents now, so they will error and then they will need to modify the `elastic-agent.yml` to set an `id` on the input definition.

The Elastic Agent does enforce that a Type/ID is unique.

cmacknz commented on x-pack/libbeat/management/managerV2.go:
```diff
@@ -251,38 +330,32 @@ func (cm *BeatV2Manager) unitListen() {
-				go cm.handleUnitReload(change.Unit)
-
+				cm.addUnit(change.Unit)
+				t.Reset(changeDebounce)
```
Can the timer be running here? If it isn't stopped the timer documentation says this isn't the correct way to handle this: https://pkg.go.dev/time#Timer.Reset

> For a Timer created with NewTimer, Reset should be invoked only on stopped or expired timers with drained channels.
> 
> If a program has already received a value from t.C, the timer is known to have expired and the channel drained, so t.Reset can be used directly. If a program has not yet received a value from t.C, however, the timer must be stopped and—if Stop reports that the timer expired before being stopped—the channel explicitly drained:
> 
> if !t.Stop() {
> 	<-t.C
> }
> t.Reset(d)
> This should not be done concurrent to other receives from the Timer's channel.

blakerouse replied:
I looked into this and it seems the way I have it in the code is correct. The reason it is correct is because the `<-t.C` is handled in the select in the same goroutine. It that was not the case it would be a problem.

By using the code as you commented above doesn't work because the `<-t.C` would then occur in multiple places and 2 places cannot read from that channel or it will cause the one that doesn't get the message from the channel to block.

Many instances of golang code use the `t.Reset` like I do in the select and seems to work as expected in the code.

cmacknz commented on x-pack/libbeat/management/managerV2.go:
```diff
@@ -311,95 +377,156 @@ func (cm *BeatV2Manager) stopBeat() {
-		return
+	if unit == nil {
+		// output is being stopped
+		err := output.Reload(nil)
```
Is there any reason why this change doesn't also trigger the Beat to restart? I'm not sure this will be any better behaved, in fact I'm not sure of the behaviour of removing the output in a Beat in general.

blakerouse replied:
Because if the output is now nil, that means beats is stopping. We would not want it to just be stopped here, once all the units have reported stopped the Elastic Agent will stop the beat.

cmacknz replied:
Makes sense, probably worth adding a clarify this.

cmacknz commented on x-pack/libbeat/management/config.go:
```diff
@@ -11,8 +11,9 @@ import (
-	Blacklist ConfigBlacklistSettings `config:"blacklist" yaml:"blacklist"`
+	Enabled       bool                    `config:"enabled" yaml:"enabled"`
+	Blacklist     ConfigBlacklistSettings `config:"blacklist" yaml:"blacklist"`
+	OutputRestart bool                    `config:"output_restart" yaml:"output_restart"`
```
nit: We used `restart_on_output_change` as the configuration and yaml name here in https://github.com/elastic/beats/pull/34049/files so that it would match exactly with the parameter name in the [v1 spec files](https://github.com/elastic/elastic-agent/blob/1fa24b7a28c0e9e30647f17d945a03168055bde5/internal/spec/filebeat.yml#L12).

I also think `restart_on_output_change` more obviously describes what this does if you read it by itself on the spec file side.

blakerouse replied:
I updated to use the same flag name `restart_on_output_change`.

cmacknz commented on x-pack/libbeat/management/managerV2_test.go:
```diff
@@ -0,0 +1,177 @@
+	require.NoError(t, err)
+	defer m.Stop()
+
+	require.Eventually(t, func() bool {
```
I think this would be a better test if we also transitioned some or all of the units expected states to stopped to ensure they are removed correctly.

It would also be nice to try to cover the logic to restart on output change but you'd have to mock out the actual stop/shutdown.

blakerouse replied:
I added more to the test to check for, state changes, log level changes, and stopping and removal of units.

cmacknz commented on x-pack/libbeat/management/managerV2_test.go:
```diff
@@ -0,0 +1,177 @@
+				},
+			},
+			State:    proto.State_HEALTHY,
+			LogLevel: proto.UnitLogLevel_INFO,
```
If you used a lower log level for one of these units you could trivially test the log level adjustment with a call to [logp.GetLevel](GetLevel).

blakerouse replied:
Added to the test to cover this.

cmacknz reviewed (commented):
A few new questions, but generally this looks good to me.

We will need a matching change to the Beat spec files to:
1. Enable the output_restart/restart_on_output_change option for each Beat.
2. Restore the default log level to INFO.

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

cmacknz commented:
@axw and @oren-zohar you will want to make sure you pick up this change to libbeat in APM server and Cloudbeat when it merges. The major highlights of what this addresses are:

1. Allow changing the log level of the `logp` logger dynamically in response to configuration changes from the Elastic agent. The previous behaviour where the agent know to restart Beats when the log level changed has been removed.
2. Fixes several potential bugs in how input configurations are reloaded dynamically, including avoiding a situation where a Beat with multiple units would incorrectly shut down when only one of them was removed.
3. Restores an old work around for dynamic output configuration reloads not working in libbeat, by automatically restarting the Beat to have the change take effect. See https://github.com/elastic/elastic-agent/issues/1913. This does not apply to APM server.

cmacknz reviewed (approved):
Looks good, thanks for the extra testing!

blakerouse commented:
/test

cmacknz commented on x-pack/libbeat/management/generate.go:
```diff
@@ -78,15 +78,33 @@ func CreateInputsFromStreams(raw *proto.UnitExpectedConfig, inputType string, ag
 		}
-		streamSource = injectGlobalProcesssors(raw, streamSource)
+
+		// now the order of the processors on this input is as follows
+		// 1. stream processors
+		// 2. agentInfo processors
+		// 3. global processors
+		// 4. stream specific processors
+
```
The order here doesn't quite line up with the order of the comments above, which is

1. Global processors
2. agentInfo processors
3. Stream processors

blakerouse replied:
It is using prepend not append. So it prepends to the array each time so the order is reversed when reading the code. Operating this way allows the processors defined on the stream to always be last.

It is definitely working as the code comment says and it is working under the Elastic Agent correctly.

cmacknz replied:
Yeah was just going to comment that I don't think I read the prepend operations correctly.

cmacknz commented:
@axw and @oren-zohar reminder ping now that this has merged that you will want to pick up this change in libbeat on both main and 8.6. 

You will also want to ensure that the elastic-agent-client version is bumped to https://github.com/elastic/elastic-agent-client/releases/tag/v7.0.3 in your go.mod files if the libbeat update does not do this automatically.

axw commented:
Thanks @cmacknz, apm-server main and 8.6 are both updated, and I've confirmed that elastic-agent-client was bumped to 7.0.3.

oren-zohar commented:
Thanks @cmacknz @blakerouse, cloudbeat was updated and I can confirm this fix works as expected :rocket:
